test/evergreen/code_coverage/per_test_code_coverage.py

```diff
@@ -0,0 +1,261 @@
+            os.mkdir(coverage_output_dir)
+            shutil.copy(src=task_info_path, dst=coverage_output_dir)
+            gcovr_command = (f"gcovr {build_copy_name} -f src -j 4 --html-self-contained --html-details "
+                             f"{coverage_output_dir}/2_coverage_report.html --json-summary-pretty "
+                             f"--json-summary {coverage_output_dir}/1_coverage_report_summary.json "
+                             f"--json {coverage_output_dir}/full_coverage_report.json")
+            split_command = gcovr_command.split()
+            env = os.environ.copy()
+            if verbose:
+                print(f'env: {env}')
+                print(f'gcovr_command: {gcovr_command}')
+            try:
+                completed_process = subprocess.run(split_command, env=env, check=True)
+                output = completed_process.stdout
+                print(f'Command returned {output}')
+            except subprocess.CalledProcessError as exception:
+                print(f'Command {exception.cmd} failed with error {exception.returncode} "{exception.output}"')
+            if verbose:
+                print(f'Completed a run of gcovr on {build_copy_name}')
+    print(f"Ending run_gcovr({build_dir_base}, {gcovr_dir})")
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-c', '--config_path', required=True, help='Path to the json config file')
+    parser.add_argument('-b', '--build_dir_base', required=True, help='Base name for the build directories')
+    parser.add_argument('-j', '--parallel', default=1, type=int, help='How many tests to run in parallel')
+    parser.add_argument('-g', '--gcovr_dir', help='Directory to store gcovr output in')
+    parser.add_argument('-s', '--setup', action="store_true",
+                        help='Perform setup actions from the config in each build directory')
+    parser.add_argument('-v', '--verbose', action="store_true", help='Be verbose')
+    args = parser.parse_args()
+
+    verbose = args.verbose
+    config_path = args.config_path
+    build_dir_base = args.build_dir_base
+    gcovr_dir = args.gcovr_dir
+    parallel_tests = args.parallel
+    setup = args.setup
+
+    if verbose:
+        print('Per-Test Code Coverage')
+        print('======================')
+        print('Configuration:')
+        print(f'  Config file                      {config_path}')
+        print(f'  Base name for build directories: {build_dir_base}')
+        print(f'  Number of parallel tests:        {parallel_tests}')
+        print(f'  Perform setup actions:           {setup}')
+        print(f'  Gcovr output directory:          {gcovr_dir}')
+        print()
+
+    if parallel_tests < 1:
+        sys.exit("Number of parallel tests must be >= 1")
+
+    # Load test config json file
+    with open(config_path) as json_file:
+        config = json.load(json_file)
+
+    if verbose:
```

Comment:
It looks as tho, we would benefit from a logger class. We pass in the verbose variable into the class which would check the if condition instead, such that whenever we do something like:
`logger.message("hello world")`, it would check the verbose variable if it should actually print to stdout or not.

Reply:
Good idea - I've introduced a logger.